Router throws an assertion failure if it has persisted data and you change the peer name

Repro: extend test 380 with the three lines at the end here:

````
# Launch two disconnected routers
weave_on $HOST1 launch --name ::1 --ipalloc-init seed=::1,::2
weave_on $HOST2 launch --name ::2 --ipalloc-init seed=::1,::2

# Ensure allocations can proceed
assert_raises "timeout 10 cat <( start_container $HOST1 --name c1)"
assert_raises "timeout 10 cat <( start_container $HOST2 --name c2)"

# Connect routers
weave_on $HOST2 connect $HOST1

# Check connectivity
assert_raises "exec_on $HOST1 c1 $PING c2"
assert_raises "exec_on $HOST2 c2 $PING c1"

# Now restart one router with a different peername
weave_on $HOST2 forget $HOST1
weave_on $HOST2 stop
weave_on $HOST2 launch --name ::3
````

Logs:

````
INFO: 2016/05/03 10:18:57.778034 Command line options: map[ipalloc-range:10.32.0.0/12 name:::3 nickname:host2 port:6783 datapath:datapath dns-effective-listen-address:172.17.0.1 dns-listen-address:172.17.0.1:53 http-addr:127.0.0.1:6784]
INFO: 2016/05/03 10:18:57.778816 Communication between peers is unencrypted.
INFO: 2016/05/03 10:18:57.779838 Our name is 00:00:00:00:00:03(host2)
INFO: 2016/05/03 10:18:57.779864 Initial set of peers: []
INFO: 2016/05/03 10:18:57.780664 Docker API on unix:///var/run/docker.sock: &[ApiVersion=1.22 GitCommit=20f81dd GoVersion=go1.5.3 Os=linux Arch=amd64 KernelVersion=4.2.0-34-generic BuildTime=2016-03-10T21:49:11.235199091+00:00 Version=1.10.3]
INFO: 2016/05/03 10:18:57.780845 [allocator 00:00:00:00:00:03] Deleting persisted data for peername 00:00:00:00:00:02
INFO: 2016/05/03 10:18:57.781955 [allocator 00:00:00:00:00:03] Initialising via deferred consensus
INFO: 2016/05/03 10:18:57.782480 Listening for DNS queries on 172.17.0.1
INFO: 2016/05/03 10:18:57.782506 Sniffing traffic on datapath (via ODP)
INFO: 2016/05/03 10:18:57.783665 Listening for HTTP control messages on 127.0.0.1:6784
panic: Assertion failure

goroutine 40 [running]:
github.com/weaveworks/weave/common.Assert(0xc821360000)
	/go/src/github.com/weaveworks/weave/common/utils.go:16 +0x6c
github.com/weaveworks/weave/ipam.(*Allocator).assertInvariants(0xc821360000)
	/go/src/github.com/weaveworks/weave/ipam/allocator.go:853 +0x146
github.com/weaveworks/weave/ipam.(*Allocator).actorLoop(0xc821360000, 0xc82134b3e0, 0xc82134b440)
	/go/src/github.com/weaveworks/weave/ipam/allocator.go:644 +0xed
created by github.com/weaveworks/weave/ipam.(*Allocator).Start
	/go/src/github.com/weaveworks/weave/ipam/allocator.go:124 +0x120
````

--- a/ipam/allocator.go
+++ b/ipam/allocator.go
@@ -113,19 +113,19 @@ func NewAllocator(config Config) *Allocator {
 
 // Start runs the allocator goroutine
 func (alloc *Allocator) Start() {
-	if alloc.loadPersistedData() {
-		if len(alloc.seed) != 0 {
-			alloc.infof("Found persisted IPAM data, ignoring supplied IPAM seed")
-		}
-	} else {
-		if len(alloc.seed) != 0 {
-			alloc.infof("Initialising with supplied IPAM seed")
-			alloc.createRing(alloc.seed)
-		} else if alloc.paxos.IsElector() {
-			alloc.infof("Initialising via deferred consensus")
-		} else {
-			alloc.infof("Initialising as observer - awaiting IPAM data from another peer")
-		}
+	loadedPersistedData := alloc.loadPersistedData()
+	switch {
+	case loadedPersistedData && len(alloc.seed) != 0:
+		alloc.infof("Found persisted IPAM data, ignoring supplied IPAM seed")
+	case loadedPersistedData:
+		alloc.infof("Initialising with persisted data")
+	case len(alloc.seed) != 0:
+		alloc.infof("Initialising with supplied IPAM seed")
+		alloc.createRing(alloc.seed)
+	case alloc.paxos.IsElector():
+		alloc.infof("Initialising via deferred consensus")
+	default:
+		alloc.infof("Initialising as observer - awaiting IPAM data from another peer")
 	}
 	actionChan := make(chan func(), mesh.ChannelSize)
 	stopChan := make(chan struct{})

--- a/ipam/allocator.go
+++ b/ipam/allocator.go
@@ -876,11 +876,13 @@ func (alloc *Allocator) loadPersistedData() {
 	if err != nil {
 		alloc.fatalf("Error loading persisted peer name: %s", err)
 	}
-	ringFound, err := alloc.db.Load(ringIdent, &alloc.ring)
+	var persistedRing *ring.Ring
+	ringFound, err := alloc.db.Load(ringIdent, &persistedRing)
 	if err != nil {
 		alloc.fatalf("Error loading persisted IPAM data: %s", err)
 	}
-	ownedFound, err := alloc.db.Load(ownedIdent, &alloc.owned)
+	var persistedOwned map[string]ownedData
+	ownedFound, err := alloc.db.Load(ownedIdent, &persistedOwned)
 	if err != nil {
 		alloc.fatalf("Error loading persisted address data: %s", err)
 	}
@@ -891,9 +893,11 @@ func (alloc *Allocator) loadPersistedData() {
 				if len(alloc.seed) != 0 {
 					alloc.infof("Found persisted IPAM data, ignoring supplied IPAM seed")
 				}
+				alloc.ring = persistedRing
 				alloc.space.UpdateRanges(alloc.ring.OwnedRanges())
 			}
 			if ownedFound {
+				alloc.owned = persistedOwned
 				for _, d := range alloc.owned {
 					for _, cidr := range d.Cidrs {
 						alloc.space.Claim(cidr.Addr)

--- a/ipam/allocator.go
+++ b/ipam/allocator.go
@@ -113,7 +113,20 @@ func NewAllocator(config Config) *Allocator {
 
 // Start runs the allocator goroutine
 func (alloc *Allocator) Start() {
-	alloc.loadPersistedData()
+	if alloc.loadPersistedData() {
+		if len(alloc.seed) != 0 {
+			alloc.infof("Found persisted IPAM data, ignoring supplied IPAM seed")
+		}
+	} else {
+		if len(alloc.seed) != 0 {
+			alloc.infof("Initialising with supplied IPAM seed")
+			alloc.createRing(alloc.seed)
+		} else if alloc.paxos.IsElector() {
+			alloc.infof("Initialising via deferred consensus")
+		} else {
+			alloc.infof("Initialising as observer - awaiting IPAM data from another peer")
+		}
+	}
 	actionChan := make(chan func(), mesh.ChannelSize)
 	stopChan := make(chan struct{})
 	alloc.actionChan = actionChan
@@ -870,7 +883,8 @@ func (alloc *Allocator) persistRing() {
 	}
 }
 
-func (alloc *Allocator) loadPersistedData() {
+// Returns true if persisted data is to be used, otherwise false
+func (alloc *Allocator) loadPersistedData() bool {
 	var checkPeerName mesh.PeerName
 	nameFound, err := alloc.db.Load(nameIdent, &checkPeerName)
 	if err != nil {
@@ -890,9 +904,6 @@ func (alloc *Allocator) loadPersistedData() {
 	if nameFound {
 		if checkPeerName == alloc.ourName {
 			if ringFound {
-				if len(alloc.seed) != 0 {
-					alloc.infof("Found persisted IPAM data, ignoring supplied IPAM seed")
-				}
 				alloc.ring = persistedRing
 				alloc.space.UpdateRanges(alloc.ring.OwnedRanges())
 			}
@@ -904,22 +915,14 @@ func (alloc *Allocator) loadPersistedData() {
 					}
 				}
 			}
-			return
+			return true
 		}
 		alloc.infof("Deleting persisted data for peername %s", checkPeerName)
 		alloc.persistRing()
 		alloc.persistOwned()
 	}
 
-	if len(alloc.seed) != 0 {
-		alloc.infof("Initialising with supplied IPAM seed")
-		alloc.createRing(alloc.seed)
-	} else if alloc.paxos.IsElector() {
-		alloc.infof("Initialising via deferred consensus")
-	} else {
-		alloc.infof("Initialising as observer - awaiting IPAM data from another peer")
-	}
-
+	return false
 }
 
 func (alloc *Allocator) persistOwned() {

--- a/ipam/allocator.go
+++ b/ipam/allocator.go
@@ -917,6 +917,11 @@ func (alloc *Allocator) loadPersistedData() bool {
 		return false
 	}
 
+	if persistedRing.Range() != alloc.universe {
+		overwritePersisted("Deleting persisted data for IPAM range %s; our range is %s", persistedRing.Range(), alloc.universe)
+		return false
+	}
+
 	alloc.ring = persistedRing
 	alloc.space.UpdateRanges(alloc.ring.OwnedRanges())
 

--- a/ipam/allocator.go
+++ b/ipam/allocator.go
@@ -901,28 +901,34 @@ func (alloc *Allocator) loadPersistedData() bool {
 		alloc.fatalf("Error loading persisted address data: %s", err)
 	}
 
-	if nameFound {
-		if checkPeerName == alloc.ourName {
-			if ringFound {
-				alloc.ring = persistedRing
-				alloc.space.UpdateRanges(alloc.ring.OwnedRanges())
-			}
-			if ownedFound {
-				alloc.owned = persistedOwned
-				for _, d := range alloc.owned {
-					for _, cidr := range d.Cidrs {
-						alloc.space.Claim(cidr.Addr)
-					}
-				}
-			}
-			return true
-		}
-		alloc.infof("Deleting persisted data for peername %s", checkPeerName)
+	overwritePersisted := func(fmt string, args ...interface{}) {
+		alloc.infof(fmt, args...)
 		alloc.persistRing()
 		alloc.persistOwned()
 	}
 
-	return false
+	if !nameFound || !ringFound {
+		overwritePersisted("No valid persisted data")
+		return false
+	}
+
+	if checkPeerName != alloc.ourName {
+		overwritePersisted("Deleting persisted data for peername %s", checkPeerName)
+		return false
+	}
+
+	alloc.ring = persistedRing
+	alloc.space.UpdateRanges(alloc.ring.OwnedRanges())
+
+	if ownedFound {
+		alloc.owned = persistedOwned
+		for _, d := range alloc.owned {
+			for _, cidr := range d.Cidrs {
+				alloc.space.Claim(cidr.Addr)
+			}
+		}
+	}
+	return true
 }
 
 func (alloc *Allocator) persistOwned() {

--- a/test/362_ipalloc_range_change.sh
+++ b/test/362_ipalloc_range_change.sh
@@ -1,0 +1,14 @@
+#! /bin/bash
+
+. ./config.sh
+
+start_suite "IPAM alloc range change"
+
+weave_on $HOST1 launch --ipalloc-range 10.2.0.0/16
+weave_on $HOST1 prime
+weave_on $HOST1 stop
+weave_on $HOST1 launch --ipalloc-range 10.3.0.0/16
+# Ensure allocations can proceed
+assert_raises "timeout 10 cat <( start_container $HOST1 )"
+
+end_suite

--- a/test/380_ipam_seed_2_test.sh
+++ b/test/380_ipam_seed_2_test.sh
@@ -19,4 +19,12 @@ weave_on $HOST2 connect $HOST1
 assert_raises "exec_on $HOST1 c1 $PING c2"
 assert_raises "exec_on $HOST2 c2 $PING c1"
 
+# Now restart one router with a different peername
+docker_on $HOST2 rm -f c2 >/dev/null
+weave_on $HOST2 forget $HOST1
+weave_on $HOST2 stop
+weave_on $HOST2 launch --name ::3
+# Check that this host has not retained the previous IPAM data
+assert "weave_on $HOST2 status ipam" ""
+
 end_suite

